cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/AbstractStorageProviderNamespaceAdmin.java

```diff
@@ -126,28 +128,70 @@ private void deleteLocation(NamespaceId namespaceId) throws IOException {
   }
 
   private void createLocation(NamespaceMeta namespaceMeta) throws IOException {
+    boolean createdHome = false;
     Location namespaceHome;
     if (hasCustomLocation(namespaceMeta)) {
-      validateCustomLocation(namespaceMeta);
-      return;
-    }
-    // no namespace custom location was provided one must be created by cdap
-    namespaceHome = namespacedLocationFactory.get(namespaceMeta);
-    if (namespaceHome.exists()) {
-      throw new FileAlreadyExistsException(namespaceHome.toString());
+      namespaceHome = validateCustomLocation(namespaceMeta);
+    } else {
+      // no namespace custom location was provided one must be created by cdap
+      namespaceHome = namespacedLocationFactory.get(namespaceMeta);
+      if (namespaceHome.exists()) {
+        throw new FileAlreadyExistsException(namespaceHome.toString());
+      }
+      // create namespace home dir
+      if (!namespaceHome.mkdirs()) {
+        throw new IOException(String.format("Error while creating home directory '%s' for namespace '%s'",
+                                            namespaceHome, namespaceMeta.getNamespaceId()));
+      }
+      createdHome = true;
     }
-    // create namespace home dir
-    if (!namespaceHome.mkdirs()) {
-      throw new IOException(String.format("Error while creating home directory '%s' for namespace '%s'",
-                                          namespaceHome, namespaceMeta.getNamespaceId()));
+    Location namespaceData = namespaceHome.append(Constants.Dataset.DEFAULT_DATA_DIR);
+    String configuredGroupName = namespaceMeta.getConfig().getGroupName();
+    boolean createdData = false;
+    try {
+      if (createdHome) {
+        // set the group id of the namespace home if configured, or the current user's primary group
+        String groupName = configuredGroupName != null
+          ? configuredGroupName : UserGroupInformation.getCurrentUser().getPrimaryGroupName();
+        namespaceHome.setGroup(groupName);
+      }
+      // create the data directory with the default permissions; then add group privileges to rwx
+      // so that all users in this group, when impersonated, can create subdirectories for their datasets
+      if (!namespaceData.mkdirs()) {
+        throw new IOException(String.format("Error while creating data directory '%s' for namespace '%s'",
+                                            namespaceData, namespaceMeta.getNamespaceId()));
+      }
+      createdData = true;
+      // the data dir should have the group from the namespace config if present, or the same group as the home dir
+      String dataGroup = configuredGroupName != null ? configuredGroupName : namespaceHome.getGroup();
+      namespaceData.setGroup(dataGroup);
+      // set the permissions to rwx for group, if a group name was configured for the namespace
+      if (configuredGroupName != null) {
```

Comment:
1. The idea is that if you don't configure a group, then you do not desire impersonation below namespace level. In that case, you also don't want the directory to be group-writable. 

2. It could be. But what would that apply to? Only to the namespace home/top-level data dir? Or to all data underneath? So in a way, the dataset would inherit the permissions from the namespace if it does not specify its own permissions? That would be a nice feature for the future.